File: app/src/main/java/app/fynnjason/fivesingletondemo/java/LazySynchronizedSingleton.java
package app.fynnjason.fivesingletondemo.java;

/**
 * Created by FynnJason on 2017/12/25.
 * 同步锁方式，在懒加载的方式上升级，保证线程安全
 */

public class LazySynchronizedSingleton {
    //第一步：私有构造
    private LazySynchronizedSingleton() {

    }

    //第二步：创建私有对象
    private static LazySynchronizedSingleton INSTANCE;

    //第三步：提供对外访问方法
    public static synchronized LazySynchronizedSingleton getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new LazySynchronizedSingleton();
        }
        return INSTANCE;
    }
}

File: app/src/main/java/app/fynnjason/fivesingletondemo/kotlin/LazySynchronizedSingleton.kt
package app.fynnjason.fivesingletondemo.kotlin

/**
 * Created by FynnJason on 2017/12/25.
 * 同步锁方式，在懒加载的方式上升级，保证线程安全
 */
class LazySynchronizedSingleton private constructor(){
    companion object {
        private var INSTANCE: LazySynchronizedSingleton? = null

        //在Kotlin中，我们就用的注释方式来加锁
        @Synchronized
        fun getInstance(): LazySynchronizedSingleton {
            if (INSTANCE == null) {
                INSTANCE = LazySynchronizedSingleton()
            }
            return INSTANCE!!
        }
    }
}